parsec/core/fs/entry_transactions.py

```diff
@@ -0,0 +1,382 @@
+    is_workspace_manifest,
+)
+from parsec.core.fs.remote_loader import RemoteLoader
+
+
+Entry = namedtuple("Entry", "access manifest")
+
+
+def from_errno(errno, message=None, filename=None, filename2=None):
+    if message is None:
+        message = os.strerror(errno)
+    return OSError(errno, message, filename, None, filename2)
+
+
+class EntryTransactions:
+    def __init__(
+        self,
+        device: LocalDevice,
+        workspace_entry: WorkspaceEntry,
+        local_storage: LocalStorage,
+        remote_loader: RemoteLoader,
+        event_bus: EventBus,
+    ):
+        self.local_author = device.device_id
+        self.workspace_entry = workspace_entry
+        self.local_storage = local_storage
+        self.remote_loader = remote_loader
+        self.event_bus = event_bus
+
+    # Right management helper
+
+    def _check_write_rights(self, path: FsPath):
+        if not self.workspace_entry.write_right:
+            raise from_errno(errno.EACCES, str(path))
+
+    # Look-up helpers
+
+    async def _get_manifest(self, access: Access) -> LocalManifest:
+        try:
+            return self.local_storage.get_manifest(access)
+        except LocalStorageMissingEntry as exc:
+            return await self.remote_loader.load_manifest(exc.access)
+
+    async def _get_entry(self, path: FsPath) -> Tuple[Access, LocalManifest]:
+        # Root access and manifest
+        assert path.parts[0] == "/"
+        access = self.workspace_entry.access
+        manifest = await self._get_manifest(access)
+        assert is_workspace_manifest(manifest)
+
+        # Follow the path
+        for name in path.parts[1:]:
+            if is_file_manifest(manifest):
+                raise from_errno(errno.ENOTDIR, filename=str(path))
+            try:
+                access = manifest.children[name]
+            except (AttributeError, KeyError):
+                raise from_errno(errno.ENOENT, filename=str(path))
+            manifest = await self._get_manifest(access)
+
+        # Return entry
+        return Entry(access, manifest)
+
+    # Locking helpers
+
+    # This logic should move to the local storage along with
+    # the remote loader. It would then be up to the local storage
+    # to download the missing manifests. This should simplify the
+    # code and help gather all the sensitive methods in the same
+    # module.
+
+    @asynccontextmanager
+    async def _lock_entry(self, path: FsPath):
+        # Load the entry
+        access, _ = await self._get_entry(path)
+
+        # Try to lock the corresponding access
+        try:
+            async with self.local_storage.lock_manifest(access) as manifest:
+                yield Entry(access, manifest)
+
+        # The entry has been deleted while we were waiting for the lock
+        except LocalStorageMissingEntry:
+            raise from_errno(errno.ENOENT, filename=str(path))
+
+    @asynccontextmanager
+    async def _lock_parent_entry(self, path):
+        # Source is root
```

Comment:
Should add a comment about why the locking strategy is parent -> children only

Reply:
No problem